File: sentimental-credit/credit.py
# Runs main
def main():
    ask = get_number()
    sum_result = half_digit(ask)
    check(sum_result, ask)


# Gets input
def get_number():
    while True:
        ask = int(input("Enter credit card number here: "))
        if ask > 0:
            return ask


# Calculate whether number is a valid card number
# Add every other number from right-left
def half_digit(ask):
    sum_result = 0
    second = False
    while ask > 0:
        end = ask % 10
        sum_result += times_two(end) if second else end
        second = not second
        ask //= 10
    return sum_result


# Multiply by 2
def times_two(end):
    multi = end * 2
    return sum(int(digit) for digit in str(multi))


# Checks if input number is a valid number
def check(sum_result, card_number):
    # Checks lenght of input number
    if sum_result % 10 == 0:
        card_str = str(card_number)
        length = len(card_str)
        # If valid lenght, determine what type of card
        if length == 15 and (int(card_str[:2]) == 34 or int(card_str[:2]) == 37):
            print("AMEX")
        elif (length == 13 and card_number // 10**12 == 4) or (length == 16 and card_number // 10**15 == 4):
            print("VISA")
        elif length == 16 and (51 <= int(card_str[:2]) <= 55):
            print("MASTERCARD")
        # If not valid card type
        else:
            print("INVALID")
    # If not valid lenght
    else:
        print("INVALID")


if __name__ == "__main__":
    main()

File: credit/credit.c
#include <cs50.h>
#include <stdio.h>

long get_number(void);
int half_digit(long ask);
int times_two(int end);
void check(int sum, long cardNumber);

int main(void)
{
    long ask = get_number();
    int sum = half_digit(ask);
    check(sum, ask);
}

// Functions

// Get user input
long get_number(void)
{
    long ask;
    do
    {
        ask = get_long("Enter credit card number here: ");
    }
    while (ask <= 0);
    return ask;
}

// Calculate whether number is a valid card number
// Add every other number from right-left
int half_digit(long ask)
{
    int sum = 0;
    bool second = false;
    while (ask > 0)
    {
        int end = ask % 10;
        sum += second ? times_two(end) : end;
        second = !second;
        ask = ask / 10;
    }
    return sum;
}

// Multiply by 2
int times_two(int end)
{
    int multi = end * 2;
    int sum = 0;
    while (multi > 0)
    {
        sum += multi % 10;
        multi = multi / 10;
    }
    return sum;
}

// Determine if input is valid
void check(int sum, long cardNumber)
{
    if (sum % 10 == 0)
    {
        // Count lenght of input
        int length = 0;
        long tempCardNumber = cardNumber;
        while (tempCardNumber > 0)
        {
            tempCardNumber /= 10;
            length++;
        }

        // ID card type from input length and first digit/s
        if (length == 15 && (cardNumber / 10000000000000 == 34 || cardNumber / 10000000000000 == 37))
        {
            printf("AMEX\n");
        }
        else if ((length == 13 && cardNumber / 1000000000000 == 4) || (length == 16 && cardNumber / 1000000000000000 == 4))
        {
            printf("VISA\n");
        }
        else if (length == 16 && (cardNumber / 100000000000000 >= 51 && cardNumber / 100000000000000 <= 55))
        {
            printf("MASTERCARD\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}
